### DISCUSSION TITLE
data loss when using websocket to transmit data

### DISCUSSION THREAD

**Discussion by Tree2015** [cesanta/mongoose#2185] [Ask questions, get answers] [answered]
I have used a total of 9 channels of mg_mgr, of which 7 are used as TCPIP Client, one is used as webserver, and the other is used as websocket. This is the premise.
Now encounter three problems:
1. The webserver and websocket are now divided into two ports. The webserver is used for the interaction of web pages and related http requests, and the websocket is used for the transmission of real-time data. Can these two ports share one? That way I don't need to forward both ports when I'm remote, which helps a lot. I used to share one port before, but when sending http requests, it will cause websocket data transmission to lose packets, and then it will be divided into two ports. Maybe I don’t know how to use mongoose correctly.
mg_mgr_init(&mgr_websocket);
mg_mgr_init(&mgr_POST);
mg_http_listen(&mgr_websocket, s_listen_on_ws, fnn_WebSocket, NULL);
mg_http_listen(&mgr_POST, s_listen_on_post, fn_POST, NULL);

2. There will be data loss in websocket transmission; in most cases, the server will transfer data from the shared memory to the client connected to mongoose, but sometimes data (some configuration commands) will be sent from the client to Mongoose, These commands are written into the shared memory by locking, and sent to the serial port by other applications; the current situation is that when the client sends data to the server, the server can receive the data normally, but occasionally affects the service For the data sending part of the end face, I don't quite understand whether I need to use multi-threading in this scenario, such as using the part of sending data as a worker thread, like (https://mongoose.ws/documentation/tutorials/multi- threaded/) described here; but I still have doubts, if the sending part is used as a worker thread, does the receiving part need to be the same?
else if (ev == MG_EV_WS_MSG)
{
// Got websocket frame. Received data is wm->data. Echo it back!
struct mg_ws_message *wm = (struct mg_ws_message *)ev_data;
// Websocket data is forwarded to the serial port
send_usart_data(wm->data.ptr, wm->data.len);
SaveLog(wm->data.ptr, wm->data.len);
}
else if (ev == MG_EV_CLOSE)
{
// Websocket connection closed
// s_websocket_ready = 0;
}
else if (ev == MG_EV_POLL)
{
if (s_websocket_ready)
{
//Forward data in real time
ProcessWebsocketData_Send(c);
}
}
3. The problem of excessive memory usage caused by uploading and downloading files in the webserver. Because the upload and download function of the file is needed in use, the size is about 100Mb. In order to improve the speed, I modified the MG_IO_SIZE to (512 * 1024). After the modification, the speed has been significantly improved, but the free command is used during the transmission process. Check the system memory, you will see that the available memory is getting smaller and smaller, and the memory occupied by the application is astonishingly greater than 6Mb. I am not sure whether this problem will cause insufficient system memory and affect the execution of other applications. I asked some other people, and they said that it does not affect much. This part of the occupied memory will be recovered by the system within a certain period of time, but I Not sure, need to reconfirm.
Before downloading the file:
free
              total used free shared buffers
Mem: 156228 147336 8892 272 8400
After downloading the file:
free
              total used free shared buffers
Mem: 156228 152768 3460 280 8960

**Comment by scaprile** [comment, marked as answer]
TL;DR + I'm sorry, I can't follow you.

Why did you do that ? One event manager can handle that

From [the docs](https://mongoose.ws/documentation/#2-minute-integration-guide)
> NOTE: Since Mongoose's core is not protected against concurrent accesses, make sure that all mg_* API functions are called from the same thread or RTOS task.

Please follow the guidelines in the [documentation](https://mongoose.ws/documentation), see the [examples](https://github.com/cesanta/mongoose/tree/master/examples) and [tutorials](https://mongoose.ws/tutorials/) available.

If you have a specific issue, open an issue and follow the issue template to give us all the necessary information to reproduce your issue.

**Reply by Tree2015** [reply]
Thank you for you reply, i think i have make a big mistake. you means that i must handle the 9 tcpip in one event manager?

**Reply by scaprile** [reply]
What is "a tcpip" ? Did you see the examples and read the tutorials ? Did you find a single example with more than one event manager ?
https://mongoose.ws/documentation/tutorials/device-dashboard/

**Reply by Tree2015** [reply]
Yes, i have do it, but havn't understand. So,thank you very much, i will read it carefully!
